### DISCUSSION TITLE
unable to create profile information to display in Nextjs

### DISCUSSION THREAD

**Discussion by amshuDhungel** [nextauthjs/next-auth#9456] [Help]
------------------------------
profile/[id]/page.tsx
-------------------------------

import ProfileAddModal from '@/Modal/ProfileAddModal'
import { TProfile } from '@/app/types';
import React from 'react'

const getPosts =  async(profileId: string):Promise<TProfile | null> => {
  try {
    const res = await fetch(`/api/profile/${profileId}`, {
      cache: 'no-store'
    });

    if (res.ok){
      const profiles = await res.json();
      return profiles
    }
  } catch (error) {
    console.log(error)
  }

  return null;
}

const ProfileImage = async({params}: {params:{profileId: string}}) => {
  const profileid = params.profileId
  const profile = await getPosts(profileid)
  return (
    <>
    <div>
    <p>Name: {profile?.profilename}</p>
      <p>Bio: {profile?.profileBio}</p>
      <p>Location: {profile?.profilelocation}</p>
      <p>Experiences: {profile?.experiences?.join(', ')}</p>
    </div>
    <div>
      <ProfileAddModal/>
    </div>
    </>
  )
}

export default ProfileImage
--------------------------------
component/ProfileForm.tsx
--------------------------------

"use client";
import { TProfile } from "@/app/types";
import React, { useState } from "react";
import { useRouter} from "next/navigation"
import { profile } from "console";

const ProfileForm = ({profile}:{profile: TProfile}) => {
  const [profilename, setProfileName] = useState("");
  const [profileBio, setProfileBio] = useState("");
  const [profilelocation, setProfileLocation] = useState("");
  const [profileimage, setProfileImage] = useState("");
  const [experiences, setExperiences] = useState<string[]>([]);
  const [experienceInput, setExperienceInput] = useState("");
  const [publicId, setPublicId] = useState("");

  const router = useRouter()

  const addExperience = (
    e: React.MouseEvent<HTMLButtonElement, MouseEvent>
  ) => {
    e.preventDefault();
    if (experienceInput.trim() !== "") {
      setExperiences((prev) => [...prev, experienceInput]);
      setExperienceInput("");
    }
  };

  const deleteExperience = (index: number) => {
    setExperiences((prev) => prev.filter((_, i) => i !== index));
  };

  const handleSubmit = async(e: React.FormEvent) =>{
    e.preventDefault();

    console.log("Submitting form....")

    try {
      const res = await fetch(`api/profile/${profile.id}`,{
        method : "POST",
        headers : {
          "Content-type": "application/json"
        },
        body: JSON.stringify({
          profilename, profileBio, profileimage, profilelocation, experiences, publicId
        }),
      });

      if(res.ok){
        const { id } = await res.json();
        console.log("Profile created successfully with ID:", id);
        router.push(`/profile/${id}`)
      }
    } catch (error) {
      console.log(error)
    }
  };

  return (
    <div className="container mx-auto">
      <h2>Profile</h2>
      <form onSubmit = {handleSubmit} className="flex flex-col gap-2">
        <input
          onChange={(e) => setProfileName(e.target.value)}
          type="text"
          placeholder="Your name"
        />
        <textarea
          onChange={(e) => setProfileBio(e.target.value)}
          placeholder="less than or equal to word bio"
          maxLength={60}
        ></textarea>
        <input
          onChange={(e) => setProfileLocation(e.target.value)}
          type="text"
          placeholder="Your Location"
        />
        {experiences &&
          experiences.map((experience, i) => (
            <div key={i} className="flex items-center gap-4">
              <span>
                <svg
                  xmlns="http://www.w3.org/2000/svg"
                  fill="none"
                  viewBox="0 0 24 24"
                  strokeWidth={1.5}
                  stroke="currentColor"
                  className="w-6 h-6"
                >
                  <path
                    strokeLinecap="round"
                    strokeLinejoin="round"
                    d="M4.26 10.147a60.438 60.438 0 0 0-.491 6.347A48.62 48.62 0 0 1 12 20.904a48.62 48.62 0 0 1 8.232-4.41 60.46 60.46 0 0 0-.491-6.347m-15.482 0a50.636 50.636 0 0 0-2.658-.813A59.906 59.906 0 0 1 12 3.493a59.903 59.903 0 0 1 10.399 5.84c-.896.248-1.783.52-2.658.814m-15.482 0A50.717 50.717 0 0 1 12 13.489a50.702 50.702 0 0 1 7.74-3.342M6.75 15a.75.75 0 1 0 0-1.5.75.75 0 0 0 0 1.5Zm0 0v-3.675A55.378 55.378 0 0 1 12 8.443m-7.007 11.55A5.981 5.981 0 0 0 6.75 15.75v-1.5"
                  />
                </svg>
              </span>
              <div className="link">{experience}</div>
              <span className="cursor-pointer" onClick = {()=> deleteExperience(i)}>
                <svg
                  xmlns="http://www.w3.org/2000/svg"
                  fill="none"
                  viewBox="0 0 24 24"
                  strokeWidth={1.5}
                  stroke="currentColor"
                  className="w-6 h-6"
                >
                  <path
                    strokeLinecap="round"
                    strokeLinejoin="round"
                    d="M14.74 9l-.346 9m-4.788 0L9.26 9m9.968-3.21c.342.052.682.107 1.022.166m-1.022-.165L18.16 19.673a2.25 2.25 0 01-2.244 2.077H8.084a2.25 2.25 0 01-2.244-2.077L4.772 5.79m14.456 0a48.108 48.108 0 00-3.478-.397m-12 .562c.34-.059.68-.114 1.022-.165m0 0a48.11 48.11 0 013.478-.397m7.5 0v-.916c0-1.18-.91-2.164-2.09-2.201a51.964 51.964 0 00-3.32 0c-1.18.037-2.09 1.022-2.09 2.201v.916m7.5 0a48.667 48.667 0 00-7.5 0"
                  />
                </svg>
              </span>
            </div>
          ))}

        <div className="flex gap-2">
          <input
            className="flex-1"
            type="text"
            onChange={(e) => setExperienceInput(e.target.value)}
            value={experienceInput}
            placeholder="Add your experiences"
          />
          <button
            onClick={addExperience}
            className="btn flex gap-2 items-center"
          >
            <span>
              <svg
                xmlns="http://www.w3.org/2000/svg"
                viewBox="0 0 20 20"
                fill="currentColor"
                className="w-5 h-5"
              >
                <path d="M10.75 4.75a.75.75 0 00-1.5 0v4.5h-4.5a.75.75 0 000 1.5h4.5v4.5a.75.75 0 001.5 0v-4.5h4.5a.75.75 0 000-1.5h-4.5v-4.5z" />
              </svg>
            </span>
            Add
          </button>
        </div>
        <button className="primary-btn" type="submit">
          Create Profile
        </button>
      </form>
    </div>
  );
};

export default ProfileForm;
---------------------
api/profile/route.ts
---------------------
import prisma from "@/lib/prismadb";
import { NextResponse } from "next/server";

export async function POST(req: Request){
    const { profilename, profileimage, profileBio, profilelocation, experiences, publicId} = await req.json();
    try {
        const newProfile = await prisma.profile.create({
            data: {
                profilename,
                profileimage,
                profileBio,
                profilelocation,
                experiences,
                publicId
            },
        });
        console.log("Profile Created");
        return NextResponse.json(newProfile)
    } catch (error) {
        console.log(error)
        return NextResponse.json({message: "Could not create profile"})
    }
}

export async function GET(){
    try {
        const profiles = await prisma.profile.findMany(
            {include: {posts :  true}}
        );
        return NextResponse.json(profiles)
    } catch (error) {
        console.log(error)
        return NextResponse.json(
            {message: "Some error occured"},
            {status: 500}
        )
        
    }
}
------------------
api/profile/[profileId]/route.ts
import prisma from "@/lib/prismadb";
import { NextResponse } from "next/server";

export async function GET(
  req: Request,
  { params }: { params: { profileId: string } }
) {
  try {
    const id = params.profileId;
    const profile = await prisma.profile.findUnique({
      where: { id },
    });
    return NextResponse.json(profile);
  } catch (error) {
    console.log(error);
    return NextResponse.json({ message: "Could not fetch profile" });
  }
}

export async function PUT(
  req: Request,
  { params }: { params: { profileId: string } }
) {
    const {profilename, profileimage, profileBio, profilelocation, experiences, publicId} = 
    await req.json();
    const id = params.profileId;

    try {
        const profile = await prisma.profile.update({
            where: {id},
            data:{
                profilename,
                profileimage,
                profileBio,
                profilelocation,
                experiences,
                publicId
            }
        });
        return NextResponse.json(profile)
    } catch (error) {
        console.log(error)
        return NextResponse.json(
            {message: "Error editing profile"},
            {status: 500}
        )
    }
}

// This is your Prisma schema file,
// learn more about it in the docs: https://pris.ly/d/prisma-schema

generator client {
  provider = "prisma-client-js"
}

datasource db {
  provider = "mongodb"
  url      = env("DATABASE_URL")
}

model Account {
  id                 String  @id @default(auto()) @map("_id") @db.ObjectId
  userId             String @db.ObjectId
  type               String
  provider           String
  providerAccountId  String
  refresh_token      String?  @db.String
  access_token       String?  @db.String
  expires_at         Int?
  token_type         String?
  scope              String?
  id_token           String?  @db.String
  session_state      String?

  user User @relation(fields: [userId], references: [id], onDelete: Cascade)

  @@unique([provider, providerAccountId])
}

model Session {
  id           String  @id @default(auto()) @map("_id") @db.ObjectId
  sessionToken String   @unique
  userId       String   @db.ObjectId
  expires      DateTime
  user         User     @relation(fields: [userId], references: [id], onDelete: Cascade)
}

model User {
  id            String  @id @default(auto()) @map("_id") @db.ObjectId
  name          String?
  email         String?   @unique
  emailVerified DateTime?
  image         String? 
  accounts      Account[]
  sessions      Session[]
  posts         Post[]

}

model VerificationToken {
  id            String  @id @default(auto()) @map("_id") @db.ObjectId
  identifier String
  token      String   @unique
  expires    DateTime

  @@unique([identifier, token])
}

model Post{
  id          String    @id @default(auto()) @map("_id") @db.ObjectId
  title       String
  content     String
  imageUrl    String?
  publicId    String?
  catName     String?
  category    Category? @relation(fields: [catName], references: [catName])
  author      User      @relation(fields: [authorEmail], references: [email])
  authorEmail String
  prices      String
  createdAt   DateTime  @default(now())
  updatedAt   DateTime  @updatedAt
  profileimage  String?
  profile     Profile? @relation(fields: [authorEmail], references: [profilename])
}

model Category {
  id      String   @id @default(auto()) @map("_id") @db.ObjectId
  catName String   @unique
  posts   Post[]
  postIDs String[] @db.ObjectId
}

model Profile {
  id             String   @id @default(auto()) @map("_id") @db.ObjectId
  publicId       String? 
  profilename    String @unique
  profileimage   String?
  profileBio     String?
  profilelocation String?
  experiences    String[]
  posts         Post[]
  postIDs       String[] @db.ObjectId
}

and this is my prisma.schema

 ✓ Compiled /api/auth/[...nextauth] in 309ms (2458 modules)
**TypeError**: Failed to parse URL from /api/profile/undefined
    at new Request (node:internal/deps/undici/undici:5272:19)
    at R (C:\Users\amsoo\OneDrive\Desktop\tiffinbox\node_modules\next\dist\compiled\next-server\app-page.runtime.dev.js:35:342062)
    ... 23 lines matching cause stack trace ...       
    at process.processTimers (node:internal/timers:514:7) {
  [cause]: TypeError: Invalid URL
      at new URL (node:internal/url:775:36)
      at new Request (node:internal/deps/undici/undici:5270:25)
      at R (C:\Users\amsoo\OneDrive\Desktop\tiffinbox\node_modules\next\dist\compiled\next-server\app-page.runtime.dev.js:35:342062)
      at doOriginalFetch (webpack-internal:///(rsc)/./node_modules/next/dist/server/lib/patch-fetch.js:383:24)
      at eval (webpack-internal:///(rsc)/./node_modules/next/dist/server/lib/patch-fetch.js:509:20)
      at eval (webpack-internal:///(rsc)/./node_modules/next/dist/server/lib/trace/tracer.js:127:36)        
      at NoopContextManager.with (webpack-internal:///(rsc)/./node_modules/next/dist/compiled/@opentelemetry/api/index.js:359:30)
      at ContextAPI.with (webpack-internal:///(rsc)/./node_modules/next/dist/compiled/@opentelemetry/api/index.js:29:58)
      at NoopTracer.startActiveSpan (webpack-internal:///(rsc)/./node_modules/next/dist/compiled/@opentelemetry/api/index.js:952:34)
      at ProxyTracer.startActiveSpan (webpack-internal:///(rsc)/./node_modules/next/dist/compiled/@opentelemetry/api/index.js:992:36)
      at eval (webpack-internal:///(rsc)/./node_modules/next/dist/server/lib/trace/tracer.js:116:103)       
      at NoopContextManager.with (webpack-internal:///(rsc)/./node_modules/next/dist/compiled/@opentelemetry/api/index.js:359:30)
      at ContextAPI.with (webpack-internal:///(rsc)/./node_modules/next/dist/compiled/@opentelemetry/api/index.js:29:58)
      at NextTracerImpl.trace (webpack-internal:///(rsc)/./node_modules/next/dist/server/lib/trace/tracer.js:116:28)
      at globalThis.fetch (webpack-internal:///(rsc)/./node_modules/next/dist/server/lib/patch-fetch.js:193:47)
      at getPosts (webpack-internal:///(rsc)/./app/profile/[id]/page.tsx:15:27)
      at ProfileImage (webpack-internal:///(rsc)/./app/profile/[id]/page.tsx:29:27)
      at Timeout._onTimeout (C:\Users\amsoo\OneDrive\Desktop\tiffinbox\node_modules\next\dist\compiled\next-server\app-page.runtime.dev.js:35:264466)
      at Timeout._onTimeout (C:\Users\amsoo\OneDrive\Desktop\tiffinbox\node_modules\next\dist\compiled\next-server\app-page.runtime.dev.js:35:264466)
      at listOnTimeout (node:internal/timers:573:17)  
      at process.processTimers (node:internal/timers:514:7) {
    code: 'ERR_INVALID_URL',
    input: '/api/profile/undefined'
  }
}

What should I do?

**Comment by db** [comment]
I'm not sure there's enough here to help out.

However, we can see that `await fetch(/api/profile/${profileId}` and that `profileId` is undefined in the error.
